yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -553,14 +560,71 @@ export class L1TxUtils {
           maxFeePerBlobGas: gasPrice.maxFeePerBlobGas!,
         })
       )?.gas;
+      this.logger?.debug('Gas used in estimateGas by blob tx', { gas: initialEstimate });
     } else {
       initialEstimate = await this.publicClient.estimateGas({ account, ...request });
+      this.logger?.debug('Gas used in estimateGas by non-blob tx', { gas: initialEstimate });
     }
 
     // Add buffer based on either fixed amount or percentage
-    const withBuffer =
-      initialEstimate + (initialEstimate * BigInt((gasConfig.gasLimitBufferPercentage || 0) * 1_00)) / 100_00n;
+    const withBuffer = this.bumpGasLimit(initialEstimate, gasConfig);
 
     return withBuffer;
   }
+
+  public async simulateGasUsed(
+    request: L1TxRequest,
+    blockOverrides: BlockOverrides<bigint, number> = {},
+    stateOverrides: StateOverride = [],
+    _gasConfig?: L1TxUtilsConfig,
+  ): Promise<bigint> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const gasPrice = await this.getGasPrice(gasConfig, false);
+
+    const nonce = await this.publicClient.getTransactionCount({ address: this.walletClient.account.address });
+
+    try {
+      const result = await this.publicClient.simulate({
+        validation: true,
+        blocks: [
+          {
+            blockOverrides,
+            stateOverrides,
+            calls: [
+              {
+                from: this.walletClient.account.address,
+                to: request.to!,
+                data: request.data,
+                maxFeePerGas: gasPrice.maxFeePerGas,
+                maxPriorityFeePerGas: gasPrice.maxPriorityFeePerGas,
+                gas: 12_000_000n,
+                nonce,
+              },
+            ],
+          },
+        ],
+      });
+      this.logger?.debug(`Gas used in simulation: ${result[0].calls[0].gasUsed}`, {
+        result,
+      });
+      if (result[0].calls[0].status === 'failure') {
+        this.logger?.error('Simulation failed', {
+          error: result[0].calls[0].error,
+        });
+        throw new Error(`Simulation failed with error: ${result[0].calls[0].error.message}`);
+      }
+      return result[0].gasUsed;
+    } catch (err) {
+      if (err instanceof MethodNotFoundRpcError || err instanceof MethodNotSupportedRpcError) {
+        // Node doesn't support simulation, return -1n gas estimate
+        this.logger?.error('Node does not support simulation API');
+        return -1n;
```

Comment:
I'd vote for rethrowing instead of using -1, in case we forget to check for that in the caller and end up propagating an invalid value.